Rework builds

soltysh opened the pull request:
This PR includes:
* [x] `Build`'s `PodName` removal #408
* [x] Adding number to builds #528, #333 
* [x] Copying labels from `Build` to build `Pod`

bparees commented on pkg/build/api/types.go:
```diff
@@ -243,6 +240,9 @@ type BuildConfig struct {
 	Triggers []BuildTriggerPolicy `json:"triggers,omitempty"`
 
+	// LatestVersion is used to inform about number of last triggered build.
+	LatestVersion int `json:"latestVersion,omitempty"`
```
s/Latest/Last/

bparees commented on pkg/build/api/v1beta1/types.go:
```diff
@@ -258,6 +255,9 @@ type BuildConfig struct {
 	Triggers []BuildTriggerPolicy `json:"triggers,omitempty"`
 
+	// LatestVersion is used to inform about number of last triggered build.
+	LatestVersion int `json:"latestVersion,omitempty"`
```
s/Latest/Last/

bparees commented on pkg/build/api/types.go:
```diff
@@ -19,9 +19,6 @@ type Build struct {
 	Message string `json:"message,omitempty"`
 
-	// PodName is the name of the pod that is used to execute the build
-	PodName string `json:"podName,omitempty"`
```
the downside to doing this is we can now never change the GetBuildPodName function since the only coupling between builds and their pods is through that function, regardless of when the build was created.  What was the motivation to remove this field?

soltysh replied:
This issue #408, but since I've added copying labels from Build to Pod, I'll also add Build.Name in Pod's labels as well (see 3rd commit). I hope that addresses your concerns.

bparees replied:
why didn't you just move PodName to be an objectReference as was described in the issue?

csrwng replied:
Clayton says:

> Builds should remove pod name, because it should be calculated without user input

bparees replied:
ok.

bparees commented on pkg/build/util/generate.go:
```diff
@@ -13,6 +14,7 @@ import (
 // GenerateBuildFromConfig creates a new build based on a given BuildConfig. Optionally a SourceRevision for the new
 // build can be specified.  Also optionally a list of image names to be substituted can be supplied.  Values in the BuildConfig
 // that have a substitution provided will be replaced in the resulting Build
+// You NEED to update BuildConfig object after calling this!
```
this is a worrisomely loose contract.

soltysh replied:
Yeah, either this or this method needs to accept some `BuildConfigUpdater` object, what do you think Ben?

bparees replied:
yeah unfortunately that's the alternative but i think it's better.

smarterclayton replied:
There is no atomic update you can do here, so you either rely on the build name being unique, or you don't try to do this.

soltysh replied:
@bparees Having thought about it a little bit more, the idea of accepting `BuildCOnfigUpdater` is bad in that if somebody update BC in the meantime I'm pushing his changes (intentionally or not). I'm proposing for this method to return new build object and it's number which then should be saved by the caller. 

@smarterclayton I'm puzzled with your suggestion, should be try to solve it or drop the idea of build naming with the incremental number at all?

bparees replied:
technically the same issue exists when we update the lasttriggeredid on the buildconfig in the imagechange_controller.  Not having transactions is awesome.

The way around it would be to break the sequenceid out of the builconfig object and have it be its own etcd key.

soltysh replied:
> technically the same issue exists when we update the lasttriggeredid on the buildconfig in the imagechange_controller. Not having transactions is awesome.

:wink: 

> The way around it would be to break the sequenceid out of the builconfig object and have it be its own etcd key.

Yeah, but... naaahhh :stuck_out_tongue_winking_eye:

bparees replied:
@soltysh I look forward to your alternative solution. :)

smarterclayton replied:
Ben and I talked about this and here's a rough summary (maybe he already conveyed):

The thing that sets LastChangeTrigger (LCT) info on the build config is the same one who would allocate the next build id.  It's ok if we skip.  We should ensure we save LCT and increment that build id at the same time, and that happens first before you create the build object (obviously).  The build identifier number can stay on the BuildConfig (it should always be the value that points to the most recent build, be monotonically incrementing, start at 1, etc).  There should be no more builds than number, but there may be less (if people delete, if crashes happen, etc).

Eventually, we will extend the "instantiate" control verb from podtemplates to buildconfig to reduce the client work.  I would not do that yet, because we want it to be consistent with upstream work.

----- Original Message -----
> > @@ -13,6 +14,7 @@ import (
> >  // GenerateBuildFromConfig creates a new build based on a given
> >  BuildConfig. Optionally a SourceRevision for the new
> >  // build can be specified.  Also optionally a list of image names to be
> >  substituted can be supplied.  Values in the BuildConfig
> >  // that have a substitution provided will be replaced in the resulting
> >  Build
> > +// You NEED to update BuildConfig object after calling this!
> 
> technically the same issue exists when we update the lasttriggeredid on the
> buildconfig in the imagechange_controller.  Not having transactions is
> awesome.
> 
> The way around it would be to break the sequenceid out of the builconfig
> object and have it be its own etcd key.
> 
> 
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1275/files#r26386026

bparees replied:
@smarterclayton not sure if i misconveyed something, but the thing that sets LCT today is the image_change_controller (which is basically just another build client like start-build and the webhook).  Builds trigged via start-build and the webhook do not go through the image_change_controller, so that is not the central place to do this currently.

smarterclayton replied:
The same "type of thing".

----- Original Message -----
> > @@ -13,6 +14,7 @@ import (
> >  // GenerateBuildFromConfig creates a new build based on a given
> >  BuildConfig. Optionally a SourceRevision for the new
> >  // build can be specified.  Also optionally a list of image names to be
> >  substituted can be supplied.  Values in the BuildConfig
> >  // that have a substitution provided will be replaced in the resulting
> >  Build
> > +// You NEED to update BuildConfig object after calling this!
> 
> @smarterclayton not sure if i misconveyed something, but the thing that sets
> LCT today is the image_change_controller (which is basically just another
> build client like start-build and the webhook).  Builds trigged via
> start-build and the webhook do not go through the image_change_controller,
> so that is not the central place to do this currently.
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1275/files#r26407709

soltysh commented:
I think I'm done for now at least, comments welcomed @bparees @csrwng 

[test]

bparees commented on pkg/build/util/generate.go:
```diff
@@ -154,3 +160,19 @@ func SubstituteImageReferences(build *buildapi.Build, oldImage string, newImage
+
+// GetNextBuildName returns name of the next build and increments BuildConfig's LastVersion.
+func GetNextBuildName(bc *buildapi.BuildConfig) string {
+	bc.LastVersion++
```
this is a race condition.  if i create multiple builds at the same time, they could all end up with the same "lastversion" value since they all start from the same BuildConfig (assuming they all did a get on buildconfig before any one of them got to the point of incrementing this value and calling Update on the BC).

you need some sort of universal locking based on the BC here.  (and even that has clustering/scaling limitations but i think we're ok there)

smarterclayton replied:
Um - if you can't globally lock, then why would you even try to do this?

soltysh replied:
Yeah, I guess @smarterclayton is right, I can guarantee synchronous execution here, but still be only inside this method, it's rather a problem that should be solved in the `GenerateBuildFromConfig` method.

kargakis replied:
> I can guarantee synchronous execution here, but still be only inside this method

It doesn't matter that the lock will be inside the method, it's pretty common.

mfojtik commented on pkg/build/controller/strategy/custom_test.go:
```diff
@@ -22,15 +24,21 @@ func TestCustomCreateBuildPod(t *testing.T) {
 
-	if actual.ObjectMeta.Name != expected.PodName {
-		t.Errorf("Expected %s, but got %s!", expected.PodName, actual.ObjectMeta.Name)
+	if e, a := buildutil.GetBuildPodName(expected), actual.ObjectMeta.Name; e != a {
```
expectedPod and actualPod ... I hate this one letter names as they are hard to parse (what is actually returned...)

bparees replied:
+5

bparees replied:
(especially since i normally assume "e" refers to an error)

soltysh replied:
Yeah, but that limits the amount of test code I need to write, but if you two insists, I'll change that.

kargakis commented on pkg/build/controller/strategy/custom.go:
```diff
@@ -40,13 +41,19 @@ func (bs *CustomBuildStrategy) CreateBuildPod(build *buildapi.Build) (*kapi.Pod,
 	}
 
+	// Need to copy the build object here so that it doesn't share pointers with
+	// the original build object, needed for modyfing labels
```
Why don't you just copy the labels of the old build into a new map and instead keep a copy of the whole build?

soltysh replied:
Fixed.

csrwng commented on pkg/build/util/generate.go:
```diff
@@ -154,3 +160,19 @@ func SubstituteImageReferences(build *buildapi.Build, oldImage string, newImage
+
+// getNextBuildNameFromBuild returns name of the next build with random uuid added at the end
+func getNextBuildNameFromBuild(build *buildapi.Build) string {
+	return fmt.Sprintf("%s-%s", build.Name, uuid.NewUUID().String())
```
you can use the generate logic upstream instead of NewUUID
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/generate.go#L36

bparees replied:
I looked into this at one point and your rest api endpoint has to explicitly implement/support that, which ours currently do not.

csrwng replied:
I just meant calling the GenerateName function directly with the original build.Name as a prefix.

smarterclayton replied:
You won't have retry until you port over to the new framework, so you're dramatically increasing the chance of a collision because the search space is smaller.

----- Original Message -----
> > @@ -154,3 +160,19 @@ func SubstituteImageReferences(build *buildapi.Build,
> > oldImage string, newImage
> >  		}
> >  	}
> >  }
> > +
> > +// GetBuildPodName returns name of the build pod.
> > +func GetBuildPodName(build *buildapi.Build) string {
> > +	return fmt.Sprintf("pod-%s", build.Name)
> > +}
> > +
> > +// GetNextBuildName returns name of the next build and increments
> > BuildConfig's LastVersion.
> > +func GetNextBuildName(bc *buildapi.BuildConfig) string {
> > +	bc.LastVersion++
> > +	return fmt.Sprintf("%s-%d", bc.Name, bc.LastVersion)
> > +}
> > +
> > +// getNextBuildNameFromBuild returns name of the next build with random
> > uuid added at the end
> > +func getNextBuildNameFromBuild(build *buildapi.Build) string {
> > +	return fmt.Sprintf("%s-%s", build.Name, uuid.NewUUID().String())
> 
> I just meant calling the GenerateName function directly with the original
> build.Name as a prefix.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1275/files#r26507050

csrwng replied:
At the very least we should shorten the build.Name so we don't end up with a name > 63 chars. afaik that issue is still open upstream.

soltysh replied:
This applies only to Build generated from Build, initially I wanted that to have some additional incremental mechanism but I failed with not enough time to do it nicely. I'll change that UUID to some timestamp and make sure we fit into 63 char limit. Thx @csrwng.

soltysh commented:
I've added first try of the new build generation logic endpoint.

soltysh commented:
Rebased, still a couple of TODO's left.

soltysh commented:
It's done. 
@smarterclayton @bparees @csrwng PTAL.

re-[test] in the mean time.

soltysh commented:
It's done & rebased.
@smarterclayton @bparees @csrwng PTAL.

re-[test] in the mean time.

bparees commented on pkg/build/api/types.go:
```diff
@@ -338,3 +341,17 @@ type BuildLog struct {
+	kapi.ObjectMeta `json:"metadata,omitempty"`
+
+	// BuildConfigName specifies the name of the build to be the source of the new build
+	BuildConfigName string `json:"buildConfigName,omitempty"`
```
why omitempty?  this shouldn't be optional should it?

csrwng replied:
You can request a build based on a buildconfig or a build

bparees replied:
yup, nm my comment.

bparees commented on pkg/build/api/validation/validation.go:
```diff
@@ -51,6 +51,17 @@ func ValidateBuildConfig(config *buildapi.BuildConfig) errs.ValidationErrorList
 
+func ValidateBuildRequest(request *buildapi.BuildRequest) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	if len(request.BuildConfigName) == 0 && len(request.BuildName) == 0 {
```
we can't generate the buildname if one isn't provided?

csrwng replied:
if you don't have a build config name, how do you know what to build?

bparees replied:
yeah i missed the use case here where you're creating a new build from an existing build.  make sense.

bparees commented on pkg/build/controller/image_change_controller.go:
```diff
@@ -81,16 +80,25 @@ func (c *ImageChangeController) HandleImageRepo(repo *imageapi.ImageRepository)
+				// This is not a retryable error. The worst case outcome of not updating the buildconfig
+				// is that we might rerun a build for the same "new" imageid change in the future,
+				// which is better than guaranteeing we run the build 2+ times by retrying it here.
+				glog.V(2).Infof("Error updating buildConfig %v: %v", config.Name, err)
```
the error is getting sent to handleError ultimately, so this is just duplicate logging.

bparees commented on pkg/client/client.go:
```diff
@@ -50,10 +51,15 @@ func (c *Client) BuildConfigs(namespace string) BuildConfigInterface {
 }
 
+// BuildGenerate provides a REST client for Build generation
+func (c *Client) BuildGenerate(namespace string) BuildGenerateInterface {
```
BuildGenerator?

soltysh replied:
The endpoint is called `generateBuild` (I think I'll rename it to `buildGenerate`) thus the same client, for consistency.

bparees replied:
1) consistency is an admirable goal, but in this case i don't think it's worth it.  A client is a noun, thus BuildGenerator makes more sense than BuildGenerate.
2) I was actually going to suggest the endpoint be renamed to buildGenerator also, but since invoking the endpoint literally generates a build, it made sense to me that the endpoint be called generateBuild.

soltysh replied:
OK, I must admit I also wanted to change it :wink:

smarterclayton replied:
Please make the endpoint /generate

smarterclayton replied:
It should be `builds/generate` using the new sub resource support.  And the method name on the client should be Generate.

bparees replied:
Method on the client being "Generate" is fine, my concern was w/ calling the interface itself "BuildGenerateInterface" (should be BuildGeneratorInterface)

smarterclayton replied:
IT should be on the Build interface.

----- Original Message -----
> >  	return newBuildLogs(c, namespace)
> >  }
> >  
> > +// BuildGenerate provides a REST client for Build generation
> > +func (c *Client) BuildGenerate(namespace string) BuildGenerateInterface {
> 
> Method on the client being "Generate" is fine, my concern was w/ calling the
> interface itself "BuildGenerateInterface" (should be
> BuildGeneratorInterface)
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1275/files#r26703237

smarterclayton replied:
Sorry, BuildConfigInterface.  See upstream clients for Binding, Status, etc.  Verbs on the core resource are direct, verbs on sub resources are "UpdateStatus".

----- Original Message -----
> IT should be on the Build interface.
> 
> ----- Original Message -----
> > >  	return newBuildLogs(c, namespace)
> > >  }
> > >  
> > > +// BuildGenerate provides a REST client for Build generation
> > > +func (c *Client) BuildGenerate(namespace string) BuildGenerateInterface
> > > {
> > 
> > Method on the client being "Generate" is fine, my concern was w/ calling
> > the
> > interface itself "BuildGenerateInterface" (should be
> > BuildGeneratorInterface)
> > 
> > 
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/openshift/origin/pull/1275/files#r26703237
>

bparees replied:
BuildConfigInterface deals with CRUD on BuildConfig objects.  giving it a generate verb seems weird.  especially since you're not generating a BuildConfig.

soltysh replied:
I was rather thinking about adding this to Build interface, thus having `Builds(ns).Generate(request)`.

soltysh commented:
@bparees @csrwng @smarterclayton finished and all green, finally. Can I please have it merged before reaching next rebase, please, please :sweat_smile:

soltysh commented on pkg/build/generator/generator.go:
```diff
@@ -0,0 +1,243 @@
+	if !kapi.ValidNamespace(ctx, &newBuild.ObjectMeta) {
+		return nil, errors.NewConflict("build", newBuild.Namespace, fmt.Errorf("Build.Namespace does not match the provided context"))
+	}
+	kapi.FillObjectMetaSystemFields(ctx, &newBuild.ObjectMeta)
```
The build created here still is vulnerable to the name length problem (<63) should I add a logic to prevent that or at least validate that?

bparees replied:
for now i'm inclined to just hope it's going to be resolved upstream soon.

bparees commented on pkg/build/controller/image_change_controller.go:
```diff
@@ -81,16 +80,23 @@ func (c *ImageChangeController) HandleImageRepo(repo *imageapi.ImageRepository)
-				return fmt.Errorf("error starting build for buildConfig %s: %v", config.Name, err)
-			}
+			// first update the config
 			if err := c.BuildConfigUpdater.Update(config); err != nil {
```
you should not be updating the BuildConfig until you've successfully created the build, otherwise you're setting the LastTriggeredImageID w/o necessarily actually having run that build and it'll never run again.

bparees commented on pkg/build/controller/image_change_controller.go:
```diff
@@ -81,16 +80,23 @@ func (c *ImageChangeController) HandleImageRepo(repo *imageapi.ImageRepository)
+			// and then trigger the generation for that config
+			request := &buildapi.BuildRequest{BuildConfigName: config.Name}
+			if _, err := c.BuildGenerator.Generate(config.Namespace, request); err != nil {
+				return ImageChangeControllerFatalError{
```
this is what i was trying to say on the scrum...  this did not used to be a fatal error, it used to be that
1) GenerateBuildFromConfig never returned an error
2) if the actual Create build REST call failed, we returned a retryable error

Now you've got it so that if Generate (which is the combination of GenerateFromBuildConfig and the Create call) fails, you throw a fatal error so it won't be retried.

bparees commented on pkg/build/generator/generator.go:
```diff
@@ -0,0 +1,243 @@
+
+// getNextBuildNameFromBuild returns name of the next build with random uuid added at the end
+func getNextBuildNameFromBuild(build *buildapi.Build) string {
+	return fmt.Sprintf("%s-%d", build.Name, int32(util.Now().Unix()))
```
if i generate a build from a build that was generated from a build, i'm going to end up with a mess.  you should strip the previous timestamp(if present) from the buildname when generating a build from a build.

soltysh replied:
Fixed.

bparees commented on pkg/build/generator/generator.go:
```diff
@@ -0,0 +1,243 @@
+		if err != nil {
+			return nil, err
+		}
+		err = g.Client.UpdateBuildConfig(ctx, bc)
```
so fundamentally you still have a race condition here because 2 Generate calls can be made simultaneously, unless etcd is going to reject the 2nd update.  And if etcd is going to reject the 2nd update, this should be a retryable failure.

(also if etcd does not reject the 2nd update, then you've also got the possibility of clobbering other updates the the BC like an update to the LastImageId or an update the user was making via CLI)

soltysh replied:
Checked it and yes etcd returns `EcodeTestFailed` when trying to update resource which was changed in the meantime (the resourceVersion has changed). For this I had to update `image_change_controller.go` logic to apply changes to newly retrieved BuildConfig object, see `updateConfig` method there. So I'm proposing to leave it as is, if user sees an error when trying to kick a build he'll have to retry it. @smarterclayton what do you think?

bparees replied:
if it's a controller that created the build, we should make sure it's being retried automatically.

bparees commented on pkg/build/webhook/controller.go:
```diff
@@ -83,14 +85,15 @@ func (c *controller) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	}
-	build, err := buildutil.GenerateBuildWithImageTag(buildCfg, revision, c.imageRepoGetter)
-	if err != nil {
+	if err := c.buildConfigUpdater.Update(buildCfg); err != nil {
```
has buildConfig been modified at this point?  why are you updating it?

bparees commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -200,6 +201,11 @@ func (d *BuildConfigDescriber) Describe(namespace, name string) (string, error)
 	return tabbedString(func(out *tabwriter.Writer) error {
 		formatMeta(out, buildConfig.ObjectMeta)
+		if buildConfig.LastVersion == 0 {
+			formatString(out, "Latest Version", "Not built")
```
s/Not/Never/

bparees commented on pkg/build/generator/generator.go:
```diff
@@ -0,0 +1,243 @@
+}
+
+type GeneratorClient interface {
+	GetBuildConfig(ctx kapi.Context, name string) (*buildapi.BuildConfig, error)
```
couldn't this just embed the BuildConfigGetter, BuildConfigUpdater, etc interfaces, rather than redefining the methods?

soltysh replied:
Methods defined in BuildConfigGetter, BuildConfigUpdater are reaching REST API endpoints, whereas above are reaching straight to etcd, see [here](https://github.com/openshift/origin/pull/1275/files#diff-3999d64b0b8fa9423b47f9feb61db2b0R216)

bparees commented on pkg/build/generator/generator.go:
```diff
@@ -0,0 +1,243 @@
+	GetImageRepository(ctx kapi.Context, name string) (*imageapi.ImageRepository, error)
+}
+
+type Client struct {
```
again why not just embed BuildConfigGetter as a field here and allow users to supply the implementation that way, since we already have those implementations?

soltysh replied:
Same as above.

bparees replied:
ok

soltysh commented:
I've reworked the endpoints according to @smarterclayton requirements. Still a couple of @bparees comments left.

soltysh commented:
Rebased on top of latest changes, which means it it uses SubResource mechanism.

soltysh commented:
@bparees I've increased the tests level for the generator package, and as discussed on IRC we're covering the retry logic already with what we heave, iow. for ImageChangeController I'm returning retry-able error, for CLI commands (start- cancel-build) I'm just returning error, and for webhook controller I'm just returning an error without any retry.

bparees commented on pkg/build/controller/image_change_controller.go:
```diff
@@ -81,18 +81,42 @@ func (c *ImageChangeController) HandleImageRepo(repo *imageapi.ImageRepository)
+// during instantiate call, it basically copies LastTriggeredImageID to fresh copy
+// of the BuildConfig object
+func (c *ImageChangeController) updateConfig(config *buildapi.BuildConfig) error {
+	item, _, err := c.BuildConfigStore.Get(config)
```
is this get just to reduce the risk of a collision on update?

I would consider handling the collision error explicitly and re-getting/updating if you get that error.  I know we sort of talked about this this morning but i don't think we described this exact retry flow.

soltysh replied:
It's not reducing risk. At this point we ALWAYS hit the collision, because BuildConfig is being updated inside the generator (LastVersion is incremented), thus we need to get fresh copy of it and copy the triggers.

bparees commented on pkg/build/controller/image_change_controller_test.go:
```diff
@@ -110,21 +111,18 @@ func TestNewImageID(t *testing.T) {
-	if buildCreator.build == nil {
-		t.Fatal("Expected new build when new image was created!")
-	}
-	if buildCreator.build.Parameters.Strategy.DockerStrategy.Image != "registry.com/namespace/imagename:newImageID123" {
```
i think you've lost something in these tests by not explicitly checking the Image field.

bparees commented on pkg/build/controller/image_change_controller_test.go:
```diff
@@ -136,18 +134,15 @@ func TestNewImageIDDefaultTag(t *testing.T) {
-	if buildCreator.build == nil {
-		t.Fatal("Expected new build when new image was created!")
-	}
-	if buildCreator.build.Parameters.Strategy.DockerStrategy.Image != "registry.com/namespace/imagename:newImageID123" {
```
same thing here.

bparees commented on pkg/build/controller/image_change_controller_test.go:
```diff
@@ -296,67 +287,51 @@ func TestNoImageIDChange(t *testing.T) {
-		t.Errorf("Image substitutions not properly setup for new build.  Expected %s, got %s |", "registry.com/namespace/imagename:newImageID123", buildCreator.build.Parameters.Strategy.DockerStrategy.Image)
-	}
-	if buildConfigUpdater.buildcfg != nil {
-		t.Fatal("Expected no buildConfig update on BuildCreate error!")
```
again it seems like you've lost part of this test...  we wannt to confirm that BuildConfig is not being updated when we get an error on build create.

bparees commented on pkg/build/generator/generator.go:
```diff
@@ -0,0 +1,262 @@
+	if err != nil {
+		return nil, err
+	}
+	if err := g.Client.UpdateBuildConfig(ctx, bc); err != nil {
```
rather than having to call update here, can you not update the buildconfig in the getNextBuildName method directly instead?  it worries me that you're relying on code that calls getNextBuildName to perform an update afterwards...i thought we agreed that was bad.

mfojtik commented on pkg/build/api/v1beta1/register.go:
```diff
@@ -11,7 +11,9 @@ func init() {
 		&BuildLog{},
+		&BuildRequest{},
 	)
+	// api.Scheme.AddKnownTypeWithName("v1beta1", "RemoteTemplate", &RemoteTemplate{})
```
nuke or explain this ;-)

mfojtik commented on pkg/build/controller/image_change_controller.go:
```diff
@@ -81,18 +81,42 @@ func (c *ImageChangeController) HandleImageRepo(repo *imageapi.ImageRepository)
+				// This is not a retryable error. The worst case outcome of not updating the buildconfig
+				// is that we might rerun a build for the same "new" imageid change in the future,
+				// which is better than guaranteeing we run the build 2+ times by retrying it here.
+				return ImageChangeControllerFatalError{
```
just a nit, why this needs to have "ImageChangeController" in name? isn't the package name appropriate namespace?

bparees replied:
there are at least two controllers in the controller package...so, no?

mfojtik commented on pkg/build/controller/strategy/docker.go:
```diff
@@ -24,9 +25,16 @@ func (bs *DockerBuildStrategy) CreateBuildPod(build *buildapi.Build) (*kapi.Pod,
+	for k, v := range build.Labels {
+		podLabels[k] = v
+	}
+	podLabels[buildapi.BuildLabel] = build.Name
```
can you move this into util.go?

mfojtik commented on pkg/build/generator/generator.go:
```diff
@@ -0,0 +1,262 @@
+	Client GeneratorClient
+}
+
+type GeneratorClient interface {
```
why not call this just Client? The package name itself is called "generator"... this would expand to: "generator.GeneratorClient"

soltysh replied:
This just just interface name, the actual implementation is already called Client, and this is what I'm using all over the code. So I'd say to leave it as is :wink:

mfojtik commented on pkg/build/generator/generator.go:
```diff
@@ -0,0 +1,262 @@
+		}
+
+		imageRepo, err := g.Client.GetImageRepository(kapi.WithNamespace(ctx, namespace), icTrigger.From.Name)
+		if err != nil {
```
err!=nil && errors.IsNotFound(err)

mfojtik commented on pkg/build/generator/generator.go:
```diff
@@ -0,0 +1,262 @@
+			return nil, err
+		}
+
+		if imageRepo == nil || imageRepo.Status.DockerImageRepository == "" {
```
len(imageRepo.Status.DockerImageRepository) == 0

mfojtik commented on pkg/build/generator/generator.go:
```diff
@@ -0,0 +1,262 @@
+		if imageRepo == nil || imageRepo.Status.DockerImageRepository == "" {
+			continue
+		}
+		glog.V(4).Infof("Found image repo %s", imageRepo.Name)
```
ImageRepository ;-)

mfojtik commented on pkg/build/generator/rest.go:
```diff
@@ -0,0 +1,57 @@
+func (s *CloneREST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, error) {
+	request, ok := obj.(*buildapi.BuildRequest)
+	if !ok {
+		return nil, fmt.Errorf("not a buildRequest: %#v", obj)
```
NewInvalid

soltysh replied:
This is k8s style of doing this.

mfojtik commented on pkg/build/util/util.go:
```diff
@@ -0,0 +1,12 @@
+
+// GetBuildPodName returns name of the build pod.
+func GetBuildPodName(build *buildapi.Build) string {
+	return fmt.Sprintf("pod-%s", build.Name)
```
return "pod-" + build.Name

no need to overuse fmt.Sprintf

mfojtik replied:
btw. does it need to have "pod-" prefix? how is that usefull? when I do osc get pods I know that all things returned are pods ;-)

bparees replied:
@mfojtik  @smarterclayton has been telling us to never use "1"+"2" and always use fmt.Sprintf.  i'm with you, though.

soltysh replied:
I'm interested in knowing the details, @smarterclayton can you elaborate please?

soltysh commented:
Rebased and addressed part of the comments. I'm in a state that it builds and passes unit tests, but just to be sure I won't loose I'm putting it here. Now I'll be working on rest of the tests and comments that I've put into code as TODO.

soltysh commented:
Rebased once again :wink:

soltysh commented:
And again :sweat:

soltysh commented:
[test] in the mean time.

soltysh commented:
@bparees @mfojtik I've addressed all your comments (unfortunately only after pushing I realized I should have done it in separate commit, for that I'm sorry). Let me know how you feel about it.

mfojtik commented:
@soltysh LGTM ship it!

soltysh commented:
[test]

mfojtik commented:
[test]

bparees commented:
lgtm. will let you do the honors on the merge tag.

soltysh commented:
re-[test] as github is having some DDoS problems

soltysh commented:
[test]

soltysh commented:
[test]

soltysh commented:
[merge]

soltysh commented:
re-[merge]

soltysh commented:
Finally!

On 8:07pm, Fri, Mar 27, 2015 OpenShift Bot <notifications@github.com> wrote:

> Merged #1275 <https://github.com/openshift/origin/pull/1275>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/1275#event-267197358>.
>
